### PR TITLE
configMigrate() returns errors + tests

### LINKED ISSUE #2659: [distributed] tests: Add tests for config migration code
https://codecov.io/gh/minio/minio/src/8acf4d112a01f3071d008ae3bc559e2ec40a6dee/cmd/config-migrate.go

For more awareness.

### PR DESCRIPTION
**Pull request by vadmeste**
Fixes #2659

### CODE REVIEW

**Comment by vadmeste** [comment]
Increase of 2.5% in code coverage is suspected. Please don't merge this.

**Comment by harshavardhana** [comment]
> Increase of 2.5% in code coverage is suspected. Please don't merge this.

Yes won't merge, add WIP in front.

**Comment by vadmeste** [comment]
@harshavardhana, I've found that tests execute fatalIf (then os.Exit()) in config.go and that should be bad for testing itself (it seems go test just exits happily without continuing to test everything)

**Comment by harshavardhana** [comment]
> @harshavardhana, I've found that tests execute fatalIf (then os.Exit()) in config.go and that should be bad for testing itself (it seems go test just exits happily without continuing to test everything)

Yes but it will only continue other non dependent tests but not the current set i believe?  - also is this patch ready for review?

**Comment by vadmeste** [comment]
> Yes but it will only continue other non dependent tests but not the current set i believe?

No. The same thing happened when I wrote a unit test for shutdown callbacks. I can confirm again with a quick code.

> also is this patch ready for review?

Yes.

File: cmd/config-migrate_test.go [hunk 1]
```diff
@@ -0,0 +1,196 @@
...
+	if err != nil {
+		t.Fatalf("Init Test config failed")
+	}
+	// remove the root folder after the test ends.
+	defer removeAll(rootPath)
+
+	setGlobalConfigPath(rootPath)
+	configPath := rootPath + "/" + globalMinioConfigFile
+
+	/* for i := 2; i <= lastConfigVersion; i++ {
```
Comment (harshavardhana) [inline cmd/config-migrate_test.go]: what is this for?

File: cmd/config.go [hunk 2]
```diff
@@ -54,7 +54,9 @@ func getConfigPath() (string, error) {
 // mustGetConfigPath must get server config path.
 func mustGetConfigPath() string {
 	configPath, err := getConfigPath()
-	fatalIf(err, "Unable to get config path.")
+	if err != nil {
```
Comment (harshavardhana) [inline cmd/config.go:57]: You should still log 'errorIf()'

File: cmd/main.go [hunk 3]
```diff
@@ -157,8 +159,13 @@ func checkMainSyntax(c *cli.Context) {
 func Main() {
 	app := registerApp()
 	app.Before = func(c *cli.Context) error {
+
+		configDir := c.GlobalString("config-dir")
+		if configDir == "" {
+			fatalIf(errors.New("config-dir is empty"), "Unable to get config file.")
```
Comment (harshavardhana) [inline cmd/main.go]: Rename this as `fatalf(errors.New("Config directory is empty"), "Unable to get config file.")`

File: cmd/config.go (hunk 2 above)
Comment (vadmeste) [reply cmd/config.go:57]: > You should still log 'errorIf()'

Currently, minio doesn't work if it fails to get config path, and main.go will show an error anyway.

**Pull request merged**